Show discriminant before overflow in diagnostic for duplicate values.

de-vri-es opened the pull request:
This PR adds the value before overflow for explicit discriminant values in the error for duplicate discriminant values. 
I found it rather confusing to see only the overflowed value.

It only does this for literals, since overflows in const evaluated arithmetic are already a hard error.

This is my first PR to the compiler, so please let me know if the implementation can be improved :)

Before:
![image](https://user-images.githubusercontent.com/786213/125850097-bf5fb7e0-d800-4386-a738-c30f41822964.png)

After:
![image](https://user-images.githubusercontent.com/786213/125850120-e2bb765d-ad86-4888-a6cb-dec34fba3fea.png)

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

de-vri-es commented on compiler/rustc_typeck/src/check/check.rs:
```diff
@@ -1497,6 +1499,25 @@ fn check_enum<'tcx>(
+            }
+        }
+    }
+    format!("{}", evaluated)
```
Note: I also removed the backticks around the number. They seemed unnecessary to me: the value is always a decimal number, which already stands out from normal text quite well. But I can revert this if desired.

jackh726 reviewed (changes requested):
So, I'd like to actually see a test where the *first* discriminant overflows (like in the OP). And I think that the backticks around the value should remain (backticks generally specify a value or code).

And finally (this one is a bit more opinionated): I feel like the exact number here isn't super helpful. It might be better to say some like "overflowed from `usize`"

de-vri-es commented:
> So, I'd like to actually see a test where the _first_ discriminant overflows (like in the OP). And I think that the backticks around the value should remain (backticks generally specify a value or code).

Makes sense. I added another test case in `src/test/ui/error-codes/E0081.rs`. Let me know if the test case should be in a new file instead, I'll move it.

I added the backticks back, and I removed the overflow value from the main error message, leaving it only in the notes. With that, it now looks like this:

![image](https://user-images.githubusercontent.com/786213/127287180-e97b2bc0-fdc3-45d8-a7c6-45bf4ac2d8ff.png)


> And finally (this one is a bit more opinionated): I feel like the exact number here isn't super helpful. It might be better to say some like "overflowed from `usize`"

Hmm, I actually like to see the value before overflow. It lets me know that neither me nor the compiler have gone mad and we're both really talking about the same number.

jackh726 commented:
Sorry for the wait here. @bors r+ rollup

bors commented:
:pushpin: Commit 4c0ff4db80d2c6cb4dc204f65f34ce2246e27bfe has been approved by `jackh726`

<!-- @bors r=jackh726 4c0ff4db80d2c6cb4dc204f65f34ce2246e27bfe -->
<!-- homu: {"type":"Approved","sha":"4c0ff4db80d2c6cb4dc204f65f34ce2246e27bfe","approver":"jackh726"} -->

de-vri-es commented:
Thanks for the review! :)
